src/badge/styled-components.js

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright (c) Uber Technologies, Inc.
+
+This source code is licensed under the MIT license found in the
+LICENSE file in the root directory of this source tree.
+*/
+// @flow
+import { styled } from '../styles/index.js';
+import type { PlacementT, ColorT, ShapeT, RoleT, OffsetT, HierarchyT } from './types.js';
+import { COLOR, SHAPE, POSITION_STYLES, PLACEMENT, HIERARCHY } from './constants.js';
+
+function getColorStyles({ $theme, $hierarchy, $color }): {|
+  color: string,
+  backgroundColor: string,
+|} {
+  const COLOR_STYLES = {
+    [HIERARCHY.primary]: {
+      [COLOR.accent]: {
+        color: $theme.colors.contentInversePrimary,
+        backgroundColor: $theme.colors.backgroundAccent,
+      },
+      [COLOR.primary]: {
+        color: $theme.colors.contentInversePrimary,
+        backgroundColor: $theme.colors.backgroundInversePrimary,
+      },
+      [COLOR.positive]: {
+        color: $theme.colors.contentInversePrimary,
+        backgroundColor: $theme.colors.backgroundPositive,
+      },
+      [COLOR.negative]: {
+        color: $theme.colors.contentInversePrimary,
+        backgroundColor: $theme.colors.backgroundNegative,
+      },
+      [COLOR.warning]: {
+        color: $theme.colors.primaryA,
+        backgroundColor: $theme.colors.backgroundWarning,
+      },
+    },
+    [HIERARCHY.secondary]: {
+      [COLOR.accent]: {
+        color: $theme.colors.contentAccent,
+        backgroundColor: $theme.colors.backgroundLightAccent,
+      },
+      [COLOR.primary]: {
+        color: $theme.colors.primaryA,
+        backgroundColor: $theme.colors.backgroundSecondary,
+      },
+      [COLOR.positive]: {
+        color: $theme.colors.contentPositive,
+        backgroundColor: $theme.colors.backgroundLightPositive,
+      },
+      [COLOR.negative]: {
+        color: $theme.colors.contentNegative,
+        backgroundColor: $theme.colors.backgroundLightNegative,
+      },
+      [COLOR.warning]: {
+        color: $theme.colors.contentWarning,
+        backgroundColor: $theme.colors.backgroundLightWarning,
+      },
+    },
+  };
+
+  return COLOR_STYLES[$hierarchy][$color];
+}
+
+export const StyledRoot = styled<{}>('div', () => {
```

Comment:
Is `Root` the best name for this component? I ask because it actually wraps around the component that the badge is positioned relative to (rather than the badge itself, which is what some might expect when seeing "root"). One alternative could be `Container`.